Add LeastSquaresProblem capabilities

jschueller opened the pull request:
- Added LeastSquaresProblem class
- Added NearestPointProblem class (OptimizationProblem(levelValue, levelFunction) is deprecated)
- Added CMinpack class which is an interface to https://github.com/devernay/cminpack, with Levenberge-Marquart algorithm

regislebrun commented on lib/src/Base/Optim/LeastSquaresProblem.cxx:
```diff
@@ -0,0 +1,110 @@
+  const QuadraticFunction sumQuad(center, constant, linear, quadratic);
+  for (UnsignedInteger i = 0; i < residualSize; ++ i)
+    quadratic(i, i, 0) = 0.5;
+  objective_ = ComposedFunction(sumQuad, residualFunction);
```
Not the most efficient way to compute half of the squared norm of a vector, but at least it exists and it works ;-)! I will add a dedicated function to compute u->||u||^2/2, its gradient u->u and its Hessian u->Id later on.

regislebrun commented on lib/src/Base/Optim/NearestPointProblem.cxx:
```diff
@@ -0,0 +1,145 @@
+  const Matrix linear(dimension_, 1);
+  const IdentityMatrix identity(dimension_);
+  const SymmetricTensor quadratic(dimension_, 1, *(identity.getImplementation().get()));
+  objective_ = QuadraticFunction(center, constant, linear, quadratic);
```
Same remark as in NonLinearLeastSquaresProblem class

regislebrun commented on python/src/LeastSquaresProblem_doc.i.in:
```diff
@@ -0,0 +1,54 @@
+
+where *B* is problem's bounds, *f* is the objective function, *g* are equality constraints, and *h* are inequality constraints.
+
+The second form represents a nearest point problem:
```
It is rather a least-squares problem: \min ||f(x)||^2/2

regislebrun commented on python/src/NearestPointProblem_doc.i.in:
```diff
@@ -0,0 +1,68 @@
+
+.. math::
+
+    \min_{x} ||x|| \\
```
Rather write \min ||x||^2/2

regislebrun commented on python/src/LeastSquaresProblem_doc.i.in:
```diff
@@ -0,0 +1,54 @@
+Examples
+--------
+>>> import openturns as ot
+>>> residualFunction = ot.SymbolicFunction(['x1', 'x2'], ['(1-x1)^2+100*(x2-x1^2)^2'])
```
Maybe a function with output dimension > 1?

jschueller commented:
@regislebrun I added a basic wrapper to cminpack in this branch with Levenberg–Marquardt so you can try it for your work on calibration

regislebrun commented:
@jschueller Good idea to postpone ROL integration, as it looks a little bit tricky to compile/install out of the box due to its large dependence set and poor internal dependence management...

regislebrun commented on lib/src/Base/Optim/CMinpack.cxx:
```diff
@@ -0,0 +1,479 @@
+}
+
+
+int CMinpack::ComputeObjective(void *p, int m, int n, const Scalar *x, Scalar *fvec, Scalar *fjac, int ldfjac, int iflag)
```
m and ldfjac are unused

regislebrun commented on lib/src/Base/Optim/CMinpack.cxx:
```diff
@@ -0,0 +1,479 @@
+  if (iflag == 1)
+  {
+    // evaluation
+    Point outP(algorithm->getProblem().getResidualFunction()(inP));
```
const Point

regislebrun commented on lib/src/Base/Optim/CMinpack.cxx:
```diff
@@ -0,0 +1,479 @@
+  else if (iflag == 2)
+  {
+    // gradient
+    Matrix gradient(algorithm->getProblem().getResidualFunction().gradient(inP).transpose());
```
lines 120 to 128 can be replaced by:
```
    const Point gradient(*algorithm->getProblem().getResidualFunction().gradient(inP).transpose().getImplementation());
    const Scalar norm = gradient.norm();
    LOGINFO(OSS() << "Gradient norm: " << norm);
    std::copy(gradient.begin(), gradient.end(), fjac);
```

jschueller commented:
@regislebrun only lmder is interesting for us so I removed the algo name parameter
I added box bound constraints support using the tanh variable change proposed in cminpack example.
I guess I'll merge it soon then.

regislebrun commented:
@jschueller I agree with you, only the nonlinear least-squares algorithm with user-provided gradient is of interest for us.
At one point we were interested into having a multivariate nonlinear equation solver. CMINPACK provides such an algorithm, but of course it is not an optimization algorithm.
IMO it is ready to merge.
